src/Chainweb/CutDB.hs

```diff
@@ -524,13 +527,56 @@ cutStreamToHeaderStream db s = S.for (go Nothing s) $ \(p, n) ->
         (HS.singleton $ LowerBound $ _blockHash p)
         (HS.singleton $ UpperBound $ _blockHash n)
 
-    o = order $ _chainGraph db
+-- | Given a stream of cuts, produce a stream of all blocks included in those
+-- cuts. Blocks of the same chain are sorted by block height.
+--
+cutStreamToHeaderDiffStream
+    :: forall m cas r
+    . MonadIO m
+    => CutDb cas
+    -> S.Stream (Of Cut) m r
+    -> S.Stream (Of (Either BlockHeader BlockHeader)) m r
+cutStreamToHeaderDiffStream db s = S.for (cutUpdates Nothing s) $ \(p, n) ->
+    S.foldrT
+        (\(cid, a, b) x -> void $ S.mergeOn toOrd x (branch cid a b))
+        (S.each $ zipCuts p n)
+  where
+    cutUpdates :: Maybe Cut -> S.Stream (Of Cut) m r -> S.Stream (Of (Cut, Cut)) m r
```

Comment:
`Stream (Of (a,b))` will cause a leak. Consider making the tuple a `T2`.

Reply:
I would the streaming steps in this function get fused into to a single loop body. So the lazy tuples shouldn't be an issue here.

I updated the PR to use `T2`, to be on the safe side.